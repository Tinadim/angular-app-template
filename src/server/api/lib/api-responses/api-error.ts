import { serverConfig } from '@server/config/server-config';

export class ApiError extends Error {
    readonly code: ErrorCode;
    readonly message: string;
    readonly errors?: ErrorDetail[];

    constructor(code: ErrorCode, message: string, errors?: ErrorDetail[]) {
        super(message);
        this.code = code;
        this.message = message;
        this.errors = errors;
    }

    toJson() {
        const errorObject: { [key: string]: any } = {
            code: this.code,
            message: this.message,
            errors: this.errors
        };

        if (serverConfig.isProduction) {
            errorObject.stack = this.stack;
        }

        return errorObject;
    }
}

export interface ErrorDetail {
    field?: string;
    code: string;
    message: string;
}

export const enum ErrorMessages {
    Forbidden = 'You don\'t have access to the requested resource',
    InvalidParameters = 'Invalid parameters provided',
    NotFound = 'Resource not found',
    Unauthorized = 'Authentication required',
    UnexpectedError = 'Unexpected error',
}

export enum ErrorCode {
    EntityNotFound = 'entityNotFound',
    Forbidden = 'accessDenied',
    InvalidParameters = 'invalidParameters',
    MissingRequiredParameter = 'missingRequiredParameter',
    Unauthorized = 'authenticationRequired',
    UnexpectedError= 'unexpectedError',
}

