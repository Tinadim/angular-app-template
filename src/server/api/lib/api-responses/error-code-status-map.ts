import { ErrorCode } from '@server/lib/exceptions/error-code';

const map: { [key in ErrorCode]: number } = {
    [ErrorCode.EntityNotFound]: 404,
    [ErrorCode.Forbidden]: 403,
    [ErrorCode.InvalidParameters]: 400,
    [ErrorCode.MissingRequiredParameter]: 400,
    [ErrorCode.Unauthorized]: 401,
    [ErrorCode.UnexpectedError]: 500,
};

export function getStatusForErrorCode(errorCode: ErrorCode) {
    return map[errorCode] || 500;
}
