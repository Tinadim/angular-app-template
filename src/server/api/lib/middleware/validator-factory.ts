import Ajv, { Options as AjvOptions } from 'ajv';
import { createValidator as createRequestValidator, SchemaCollection, ValidatorOptions } from 'express-ajv';

export const defaultAjvOptions: AjvOptions = {
    allErrors: true,
    coerceTypes: true,
    format: 'full',
    useDefaults: true,
};

const ajv = new Ajv(defaultAjvOptions);
const options: ValidatorOptions = { ajv };

export function createValidator(schemas: SchemaCollection) {
    return createRequestValidator(schemas, options);
}
