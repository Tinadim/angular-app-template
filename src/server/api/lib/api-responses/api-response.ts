import * as HttpStatusCodes from 'http-status-codes';
import { Response } from 'express';
import { ApiError, ErrorCode, ErrorDetail, ErrorMessages } from './api-error';

export class ApiResponse {
    readonly status: number;
    readonly success: boolean;
    readonly data?: object;
    readonly error?: ApiError;

    constructor(status: number, data?: object, error?: ApiError) {
        this.status = status;
        this.data = data;
        this.error = error;
        this.success = this.isSuccessResponse();
    }

    static Ok(data?: object) {
        return new ApiResponse(HttpStatusCodes.OK, data);
    }

    static Created() {
        return new ApiResponse(HttpStatusCodes.CREATED);
    }

    static NoContent() {
        return new ApiResponse(HttpStatusCodes.NO_CONTENT);
    }

    static BadRequest(error: ApiError) {
        return new ApiResponse(HttpStatusCodes.BAD_REQUEST, undefined, error);
    }

    static Unauthorized() {
        const error = new ApiError(ErrorCode.Unauthorized, ErrorMessages.Unauthorized);
        return new ApiResponse(HttpStatusCodes.UNAUTHORIZED, undefined, error);
    }

    static Forbidden() {
        const error = new ApiError(ErrorCode.Forbidden, ErrorMessages.Forbidden);
        return new ApiResponse(HttpStatusCodes.FORBIDDEN, undefined, error);
    }

    static NotFound(message?: string) {
        const error = new ApiError(ErrorCode.EntityNotFound, message || ErrorMessages.NotFound);
        return new ApiResponse(HttpStatusCodes.NOT_FOUND, undefined, error);
    }

    static TooManyRequests() {
        return new ApiResponse(HttpStatusCodes.TOO_MANY_REQUESTS);
    }

    static ServerError(error: any) {
        const errors: ErrorDetail[] = [{
            code: ErrorCode.UnexpectedError,
            message: error.message,
            field: 'originalError',
        }];
        const serverError = new ApiError(ErrorCode.UnexpectedError, ErrorMessages.UnexpectedError, errors);
        return new ApiResponse(HttpStatusCodes.INTERNAL_SERVER_ERROR, undefined, serverError);
    }

    private isSuccessResponse() {
        return this.status < 400;
    }

    toJson() {
        return {
            data: this.data,
            error: this.error.toJson(),
        };
    }

    send(res: Response) {
        res.status(this.status).json(this.toJson());
    }
}
