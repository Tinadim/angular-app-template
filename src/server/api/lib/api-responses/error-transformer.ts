import ajv, { ErrorObject } from 'ajv';
import { get, map } from 'lodash';
import { ApiError, ErrorCode, ErrorDetail, ErrorMessages } from './api-error';
import { ApiResponse } from './api-response';

const keywordMap: { [key: string]: ErrorCode | undefined } = {
    required: ErrorCode.MissingRequiredParameter,
};

function mapAjvError(ajvError: ErrorObject): ErrorDetail {
    let { message } = ajvError;
    const code = keywordMap[ajvError.keyword] || ErrorCode.InvalidParameters;
    const params = get(ajvError, 'params', {}) as any;
    const fieldName = (ajvError.dataPath || params.property || params.missingProperty ||
        params.additionalProperties || '').replace('.', '');
    if (ajvError.keyword === 'enum') {
        const allowedValues = get(ajvError, 'params.allowedValues');
        message += `: ${allowedValues}`;
    }

    return {
        code,
        message,
        field: fieldName,
    };
}

export function fromAjvError(ajvError: ajv.ValidationError): ApiResponse {
    const apiError = new ApiError(
        ErrorCode.InvalidParameters,
        ErrorMessages.InvalidParameters,
        map(ajvError.errors, mapAjvError)
    );
    return ApiResponse.BadRequest(apiError);
}
