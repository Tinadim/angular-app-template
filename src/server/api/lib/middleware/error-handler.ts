import ajv from 'ajv';
import { NextFunction, Request, Response } from 'express';
import { ApiResponse } from '../api-responses/api-response';
import { ApplicationError } from '@server/lib/exceptions/application-error';
import { mapAjvError } from '@server/lib/exceptions/error-adapters';
import { unexpectedError } from '@server/lib/exceptions/application-exceptions';

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export function errorHandler(error: any, req: Request, res: Response, next: NextFunction) {
    let mappedError: ApplicationError;

    if (error instanceof ApplicationError) {
        mappedError = error;
    } else if (error instanceof ajv.ValidationError) {
        mappedError = mapAjvError(error);
    } else {
        mappedError = unexpectedError(error);
    }

    ApiResponse.fromApplicationError(mappedError).send(res);
}
