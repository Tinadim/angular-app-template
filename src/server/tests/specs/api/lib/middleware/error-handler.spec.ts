import Ajv from 'ajv';
import { errorHandler } from '@server/api/lib/middleware/error-handler';
import { ApiResponse } from '@server/api/lib/api-responses/api-response';

describe('error-handler.spec.ts', () => {
    let mockResponse: any;

    beforeEach(() => {
        mockResponse = {
            status: jest.fn().mockImplementation(function() {
                return this;
            }),
            json: jest.fn()
        };
    });

    function assertError(expectedStatus: number) {
        expect(mockResponse.status).toHaveBeenCalledWith(expectedStatus);
        expect(mockResponse.json).toHaveBeenCalledTimes(1);
        const response = mockResponse.json.mock.calls[0][0];
        expect(response).toMatchSnapshot();
    }

    it('Should handle api-responses instances as errors', () => {
        const apiResponse = ApiResponse.NotFound('Entity 1234 was not found');
        errorHandler(apiResponse, {} as any, mockResponse, () => {});
        expect(mockResponse.status).toHaveBeenCalledWith(apiResponse.status);
        expect(mockResponse.json).toHaveBeenCalledWith(apiResponse.toJson());
    });

    it('Should handle ajv-errors instances and return the appropriate mapped error', () => {
        const ajv = new Ajv({ allErrors: true, format: 'full' });
        const schema = {
            type: 'object',
            properties: {
                name: {
                    type: 'string',
                    minLength: 3
                },
                age: {
                    type: 'number',
                    minimum: 1
                },
                city: {
                    type: 'string'
                }
            },
            required: ['name', 'age', 'city']
        };
        ajv.validate(schema, { name: 'a', age: 0 });
        const error = new Ajv.ValidationError(ajv.errors);
        errorHandler(error, {} as any, mockResponse, () => {});
        assertError(400);
    });

    it('Should handle generic errors and return the appropriate mapped error', () => {
        const error = new Error('Some random error');
        errorHandler(error, {} as any, mockResponse, () => {});
        assertError(500);
    });
});
