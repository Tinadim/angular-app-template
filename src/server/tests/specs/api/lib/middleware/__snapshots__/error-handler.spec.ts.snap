// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error-handler.spec.ts Should handle ajv-errors instances and return the appropriate mapped error 1`] = `
Object {
  "data": undefined,
  "error": Object {
    "code": "invalidParameters",
    "errors": Array [
      Object {
        "code": "invalidParameters",
        "field": "name",
        "message": "should NOT be shorter than 3 characters",
      },
      Object {
        "code": "invalidParameters",
        "field": "age",
        "message": "should be >= 1",
      },
      Object {
        "code": "missingRequiredParameter",
        "field": "city",
        "message": "should have required property 'city'",
      },
    ],
    "message": "Invalid parameters provided",
  },
}
`;

exports[`error-handler.spec.ts Should handle generic errors and return the appropriate mapped error 1`] = `
Object {
  "data": undefined,
  "error": Object {
    "code": "unexpectedError",
    "errors": Array [
      Object {
        "code": "unexpectedError",
        "field": "originalError",
        "message": "Some random error",
      },
    ],
    "message": "Unexpected error",
  },
}
`;
